### PR TITLE
Add pooling support for 3d channels last

### PR DESCRIPTION
**Pull request by dmenig**
Part of a multi-PR work to improve #59168

Meant to complete 
Write native kernels for AvgPool3d
Write native kernels for MaxPool3d
Write native kernels for AdaptiveAvgPool3d
Write native kernels for AdaptiveMaxPool3d

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @albanD

### CODE REVIEW

**Comment by hyperfraise** [comment]
I think I need approval to keep debugging with the CI please

**Comment by ezyang** [comment]
I added you to the repo, that should durably give you rights

**Comment by hyperfraise** [comment]
Thx, this is my current error in the CI, I don't get it. Can anyone provide some insight please ?
```distributed/test_store 1/1 failed! Received signal: SIGIOT
Emiting td_test_failure_stats
could not find path /var/lib/jenkins/workspace/.additional_ci_files/previous_failures.json
Traceback (most recent call last):
  File "/var/lib/jenkins/workspace/test/run_test.py", line 1812, in <module>
    main()
  File "/var/lib/jenkins/workspace/test/run_test.py", line 1764, in main
    run_tests(
  File "/var/lib/jenkins/workspace/test/run_test.py", line 1620, in run_tests
    raise RuntimeError(failure.message)
RuntimeError: distributed/test_store 1/1 failed! Received signal: SIGIOT```

**Comment by ezyang** [comment]
It might just be flaky master, check what dr ci says (i can't see it anymore since you repushed)

**Comment by hyperfraise** [comment]
This now passes. Please review :)

**Comment by ezyang** [comment]
@mingfeima will you be reviewing this?

**Comment by ppwwyyxx** [comment]
btw, don't say `fix #59168` in the PR description if you don't intend to let the PR close the issue

**Comment by hyperfraise** [comment]
I still think this is relevant.

**Comment by ezyang** [comment]
Let's fine another reviewer. Hmm...

File: aten/src/ATen/native/AdaptiveAveragePooling3d.cpp [hunk 1]
```diff
@@ -317,6 +317,12 @@ Tensor adaptive_avg_pool3d_symint(Tensor const& input, SymIntArrayRef output_siz
     // in this case, adaptive pooling is just computing mean over hw
     // dimensions, which can be done more efficiently
     Tensor out = input.mean({-1, -2, -3}, /* keepdim = */ true);
+    if (input.suggest_memory_format() == at::MemoryFormat::ChannelsLast3d) {
+      // assert ndim == 5, since ndim = 4 doesn't give channels_last
+      const auto n = input.sym_size(0);
+      const auto c = input.sym_size(1);
+      out.as_strided__symint({n, c, 1, 1, 1}, {c, 1, c, c, c});
```
Comment (ezyang) [inline aten/src/ATen/native/AdaptiveAveragePooling3d.cpp:324]: If I understand correctly, this restride doesn't change the semantic content of the tensor, but mucks around the strides for the size 1 dimensions so it looks channels last 3d.

File: aten/src/ATen/native/cpu/AvgPoolKernel.cpp [hunk 2]
```diff
@@ -531,13 +531,596 @@ void avg_pool2d_backward_kernel_impl(
...
+      break;
+    }
+    default:
+      TORCH_CHECK(false, "Unsupported memory format. Supports only ChannelsLast, Contiguous");
+  }
+}
+
+
+template <typename scalar_t>
+void cpu_avg_pool3d(
```
Comment (ezyang) [inline aten/src/ATen/native/cpu/AvgPoolKernel.cpp:551]: What do you think a good code review strategy for these kernels are? Are they similar to their non-channels last brethren and I just need to diff them, or is there more to it?

File: aten/src/ATen/native/AdaptiveAveragePooling3d.cpp (hunk 1 above)
Comment (hyperfraise) [reply aten/src/ATen/native/AdaptiveAveragePooling3d.cpp:324]: Damn @ezyang I wish I could answer you. I jsut copied and adapted [this code](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/AdaptiveAveragePooling.cpp#L129-135) the best I could and hoped someone that knows more than me could confirm I didn't do too bad.

File: aten/src/ATen/native/cpu/AvgPoolKernel.cpp (hunk 2 above)
Comment (hyperfraise) [reply aten/src/ATen/native/cpu/AvgPoolKernel.cpp:551]: This is purely copied and adapted code of `cpu_avg_pool`, `cpu_avg_pool_channels_last`, which are the 2d counterparts. I wrote this by copying those functions then writing the 3d part of it. Thus I believe you can diff those with the previous version of those function, or even diff those with the renamed 2d version.

Comment (hyperfraise) [reply aten/src/ATen/native/cpu/AvgPoolKernel.cpp:551]: Thank you for giving this your time !

**Review by ezyang — APPROVED** [review]

**Comment by hyperfraise** [comment]
I guess those are user facing, no ? @ezyang

**Pull request closed without merging**